<a href="https://colab.research.google.com/github/Fabricio151205/Machine-Learning-/blob/main/Entrenamiento_del_proyecto_de_Machine_Learning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Importar Librerias y Dataset**

In [ ]:
import pandas as pd

data = pd.read_csv('ventas_ia.csv')

print(data.head())
print()
print(data.info())
print()
print(data.describe())

   producto   año  mes  cantidad_vendida
0  Lapicero  2023    1                45
1  Lapicero  2023    2                58
2  Lapicero  2023    3                83
3  Lapicero  2023    4                41
4  Lapicero  2023    5                48

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1050 entries, 0 to 1049
Data columns (total 4 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   producto          1050 non-null   object
 1   año               1050 non-null   int64 
 2   mes               1050 non-null   int64 
 3   cantidad_vendida  1050 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 32.9+ KB
None

               año          mes  cantidad_vendida
count  1050.000000  1050.000000       1050.000000
mean   2024.285714     6.071429         21.465714
std       1.030648     3.426978         19.498754
min    2023.000000     1.000000          0.000000
25%    2023.000000     3.000000          6.000000
50%    2024.00

In [ ]:
from sklearn.preprocessing import LabelEncoder

encoder = LabelEncoder()

data['producto_id'] = encoder.fit_transform(data['producto'])

print(data[['producto', 'producto_id']].drop_duplicates().head(20))

            producto  producto_id
0           Lapicero           11
42       Cuaderno A4            6
84             Lápiz           12
126         Borrador            1
168          Tajador           21
210         Plumones           18
252        Corrector            5
294       Resaltador           19
336       Papel Bond           14
378           Folder            8
420        Cartulina            3
462         Silicona           20
504           Tijera           23
546   Cinta Adhesiva            4
588        Grapadora            9
630           Grapas           10
672       Archivador            0
714         Pinceles           17
756         Temperas           22
798  Mochila Escolar           13


In [ ]:
X = data[['producto_id', 'año', 'mes']]
y = data['cantidad_vendida']

print(X.head())
print()
print(y.head())

   producto_id   año  mes
0           11  2023    1
1           11  2023    2
2           11  2023    3
3           11  2023    4
4           11  2023    5

0    45
1    58
2    83
3    41
4    48
Name: cantidad_vendida, dtype: int64


##**Entrenamiento de la IA**

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

X = data[['producto_id', 'año', 'mes']]
y = data['cantidad_vendida']

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

modelo = LinearRegression()

modelo.fit(X_train, y_train)

print("Modelo entrenado correctamente")

Modelo entrenado correctamente


##**Pruebas de que la IA se entreno**

In [ ]:
prediccion = modelo.predict([[11, 2027, 1]])

print(prediccion)

[33.18078916]


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


##**Medir Prediccion**

In [ ]:
from sklearn.metrics import mean_absolute_error, r2_score

predicciones = modelo.predict(X_test)

mae = mean_absolute_error(y_test, predicciones)

r2 = r2_score(y_test, predicciones)

print("MAE:", mae)
print("R2:", r2)

MAE: 15.104924780540513
R2: 0.020876393099720225


**MAE** es el error promedio.

---


Ejemplo:

> "MAE = 3"
eso significa que la IA se equivoca aproximadamente en 3 unidades.


---

**R2** es un medidor de calida

---
Va de 0 a 1, mientras mas cerca del 1 es mejor



*   0.30 = malo
*   0.50 = aceptable
*   0.70 = bueno
*   0.80 = muy bueno
*   0.90 = excelente


##**Segunda Prueba**

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, r2_score

X = data[['producto_id', 'año', 'mes']]
y = data['cantidad_vendida']

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42
)

modelo_rf = RandomForestRegressor(
    n_estimators=200,
    random_state=42
)

modelo_rf.fit(X_train, y_train)

predicciones = modelo_rf.predict(X_test)

mae = mean_absolute_error(y_test, predicciones)
r2 = r2_score(y_test, predicciones)

print("MAE:", mae)
print("R2:", r2)

MAE: 2.8675476190476195
R2: 0.948491524997847


#**Explicacion**
1.- En la Primera prueba concluimos que los datos ingresador en el dataset arrojaban un porcentaje muy bajo de presición porque usamos **LinearRegression()**

Pero las ventas de una papelería tienen:

*  Estacionalidad
*  Diferencias entre productos
*  Variaciones no lineales

y una regresión lineal simple no captura eso muy bien.

2.- En la Segunda prueba concluimos que los datos ingresados arrojaron un porcentaje mucho mejor al inicial y eso se debe a que usamos **RandomForestRegressor**

Porque:

*  Funciona mejor con datos tabulares.
*  No exige relaciones lineales.
*  Suele dar mejores resultados para predicción de demanda.

##**Guardar modelo entrenado**

In [ ]:
import joblib

joblib.dump(modelo_rf, 'modelo_demanda.pkl')
joblib.dump(encoder, 'encoder_productos.pkl')

['encoder_productos.pkl']

#**Conclusión**
1. Para la predicción de demanda se evaluaron diferentes algoritmos de Machine
Learning. Inicialmente se utilizó Regresión Lineal; sin embargo, los resultados obtenidos presentaron una baja capacidad predictiva debido a la naturaleza no lineal de los datos de ventas.

2. Posteriormente se implementó el algoritmo Random Forest Regressor, obteniendo un Error Absoluto Medio (MAE) de 2.87 unidades y un coeficiente de determinación (R²) de 0.948. Esto indica que el modelo es capaz de explicar aproximadamente el 95% de la variabilidad de las ventas históricas, logrando una alta precisión en la estimación de la demanda futura de productos de papelería.
